### PR TITLE
feat: dark mode

### LINKED ISSUE #607: Dark mode/theme
### Is your feature request related to a problem? Please describe.
A dark color scheme for the project's web view, loved by a lot of devs. I think this project would really deserve something like this for the open-source night sessions. 

### Describe the solution you'd like
A new dark theme, with a switch to be able to choose from a light (current) or the new dark theme. It could be changed by the [Color Scheme Preference API](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) or with a manual rewrite. A customized [dark-mode-toggle](https://www.npmjs.com/package/dark-mode-toggle) would be an easy to implement solution, that could be customized to the project's needs.

### PR DESCRIPTION
**Pull request by mtfoley**
## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description
PR to fix #607 by adding dark mode support.  Tested a handful of things so far in chrome, but plenty of work to do.  Will try to get out of draft next week if all looks in the right direction.

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 docs
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
Make sure it looks nice 👍 

## [optional] What gif best describes this PR or how it makes you feel?
cruella

### CODE REVIEW

**Comment by mtfoley** [comment]
@bdougie @filiptronicek @0-vortex I've got this draft PR started and I definitely have some more tweaks to make - icons, some muted text, etc. please take a look at the Netlify deploy previews and let me know what y'all feel would drive to this being ready to pull out of draft.

**Review by filiptronicek — COMMENTED** [review]
I'm really glad this is being worked on. I like the changes so far, maybe I would make the `#0C0E0E` which is used for many elements a lighter grey, but it's looking good.

We could also add a user preference switch for people who prefer either design independent of their set system theme. Like so:
image

**Comment by mtfoley** [comment]
@filiptronicek thx for taking a look. Most of my changes so far have been adjusting adding CSS with a media selector. In order give the user some kind of on demand selection, I think I would need to backtrack these to use CSS that targets based on the body element class and have JS that adjusts that class, instead of using media selector... does that sound right?

**Comment by filiptronicek** [comment]
> @filiptronicek thx for taking a look. Most of my changes so far have been adjusting adding CSS with a media selector. In order give the user some kind of on demand selection, I think I would need to backtrack these to use CSS that targets based on the body element class and have JS that adjusts that class, instead of using media selector... does that sound right?

Yes, that's about how I feel like it can be done. We could make it so the js adds a class when the page loads and also adds an event listener to change the class even when the user changes the color scheme preference when the page is already loaded.

**Comment by mtfoley** [comment]
Ok, I don't have any past experience with the styled-components but I'll see what it lets me do :-) any prior used components in the project that you'd suggest for mode choosing, and where to place it?

**Comment by filiptronicek** [comment]
> 
> 
> Ok, I don't have any past experience with the styled-components but I'll see what it lets me do :-) any prior used components in the project that you'd suggest for mode choosing, and where to place it?

@mtfoley although this modal was not done in React, here is the minimal version of this, which saves the preference to the local storage: https://codepen.io/filip_tronicek/pen/RwpxqKo

**Comment by bdougie** [comment]
Sorry that comment may be pre-mmature. I am interested in this `prefers-color-scheme: dark`, I have not used or heard of that.

**Comment by mtfoley** [comment]
In order to test it I had to go into Chrome Developer Tools / Rendering, and I think the setting is "Emulate CSS prefers-color-scheme: dark" or something to that effect. On my iPhone the deploy preview shows up in dark mode because that's what I have picked in display settings. On my MacBook it changes with the time of day, I think because of the dynamic desktop photo thing.

**Comment by bdougie** [comment]
> In order to test it I had to go into Chrome Developer Tools / Rendering, and I think the setting is "Emulate CSS prefers-color-scheme: dark" or something to that effect. On my iPhone the deploy preview shows up in dark mode because that's what I have picked in display settings. On my MacBook it changes with the time of day, I think because of the dynamic desktop photo thing.

This might need some documentation included with the PR. I was not familiar with all that myself.

**Comment by 0-vortex** [comment]
## Some screenshots from Big Sur 11.4 Beta (20F5055c)

### Safari 14.1.1 (16611.2.5):

Screenshot 2021-06-01 at 21 13 43

### Chrome 92.0.4515.20 (Official Build) dev (x86_64):

Screenshot 2021-06-01 at 21 15 54

The color scheme seems all good without checking any experiments, the only thing I am noticing is different font x-width and unrelated to this PR, the use-shopping-cart could use some re-thinking.

**Comment by mtfoley** [comment]
Ok, after reading all the commentary I think I'm going to take a step back and sort of start over with this. I think I have some reading to do on this article: https://dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0

To me, the behavior scope sounds like:
- app detect user's current `prefers-color-scheme` value, as well as any overrides in local storage.
- app adjust the body or html element class to reflect the active color scheme and let CSS take over from that point

And the implementation scope sounds like it needs to include some documentation as well for this.

I'm not sure how this approach will work with the styled-components, but we'll find out soon enough.

**Comment by 0-vortex** [comment]
> Ok, after reading all the commentary I think I'm going to take a step back and sort of start over with this. I think I have some reading to do on this article: [dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0](https://dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0)
> 
> To me, the behavior scope sounds like:
> 
> * app detect user's current `prefers-color-scheme` value, as well as any overrides in local storage.
> * app adjust the body or html element class to reflect the active color scheme and let CSS take over from that point
> 
> And the implementation scope sounds like it needs to include some documentation as well for this.
> 
> I'm not sure how this approach will work with the styled-components, but we'll find out soon enough.

I feel like that article is needlessly complicated, trying to cover space instead of simplifying what it's doing and instead go for a "tiered" learning approach:
- styled-components docs: https://styled-components.com/docs/advanced#via-usecontext-react-hook
- dark mode mental framework: https://css-tricks.com/dark-modes-with-css/
- styled-components demo: https://codesandbox.io/s/3ljgt?file=/src/index.js 
- building a react-dark-toggle with styled-components: https://www.smashingmagazine.com/2020/04/dark-mode-react-apps-styled-components/ - would replace your linked tutorial with this one
- and a common used package for anything but the styling itself: https://www.npmjs.com/package/react-toggle-dark-mode

Not being too into styled-components I would split 3 ways, toggle component, set up `<ThemeProvider>` to check local storage for `ThemeMode ` value and create a function to manipulate `ThemeMode`. Then I would able to use that 2 way switch anywhere by just reading the styled-components theme from the react hooks. This is how that split would roughly look like, this could be `Header` component leveraging `react-toggle-dark-mode`:

```jsx
import { ThemeContext } from 'styled-components';
const themeContext = useContext(ThemeContext);

<DarkModeSwitch
	  checked={themeContext === 'light'}
	  onChange={() => {
	    setDarkMode(themeContext === 'light' ? 'light' : 'dark');
	  }}
	  size={42}
/>
```

Hope this removes some confusion, learning 3 things at once you already got 2 of them right 🤩

**Comment by mtfoley** [comment]
Well that was fun - I feel like I'm getting close now.
I've got a ThemeContext and I'm differentiating between dark/light/system, falling back to system.
I've got an unstyled triple button that lets you update the theme when logged in.
I changed my media queries to use `body.class & { }` instead.

Known issues:
- the buttons need to look nice. Maybe have icons like in some examples I've seen. Not sure what icon to use for "system" or "automatic".
- the buttons don't adjust the theme when not logged in... need to take some time with that (also, should we even show them?)
- I think I've still clobbered the background of the search field in AddRepoForm. I need to look back at the original style of it in main branch.
- I need to add docs.

**Comment by notBdougie** [comment]
Perhaps this HEX instead of black #241E1D

**Comment by bdougie** [comment]
I switched my phone to dark mode to get a feel fork dark mode in different apps and like twitters lines in between tweets. I think what you have so far is really good. 

I don't think you really need to have it. Lose to the original hue at all. Disregard my comment on stream. 

The buttons to switch it on and off is a bonus.

image

**Review by mtfoley — COMMENTED** [review]
don't know how to turn this review off...

**Comment by mtfoley** [comment]
@bdougie @0-vortex @filiptronicek I think we are within striking distance. 

As a bonus, this PR deletes that typo where the word "repo" ended up next to the repo language labels :-)

**Review by filiptronicek — CHANGES_REQUESTED** [review]
This PR is indeed very close to perfection, I just have a couple things:

- The icons may look a lot better if they were white in all theme settings (we can just use [`filter: invert()`](https://developer.mozilla.org/en-US/docs/Web/CSS/filter-function/invert()) for this.)
- The skeleton loading is still in light colors, and that looks a bit off
- Maybe add a bit of a margin to the theme selection? 

image
image

**Comment by mtfoley** [comment]
Alright I'll take a look at these. The octicons render svg straight to the DOM so I think those just inherit fill by the font color. 

For the skeleton stuff I think I may need to play with className attributes since it's been mostly CSS approach so far and for the most part the theme context doesn't have a lot of effect on the components in the tree. I think the more neutral grey used in some of these can probably stay as is, it's the card background that needs to match the hydrated state.

**Comment by bdougie** [comment]
I personally think the svgs can stay as is until a solution is discovered. I think we can open a good first issue for that honestly. I can take a closer look at everything when I am back at a keyboard.
<details><summary>Browser metadata</summary>





```
Path:      /
Browser:   Mobile Safari 14.1.1 on iOS 14.6
Viewport:  375 x 635 @3x
Language:  en-us
Cookies:   Disabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-1020--open-sauced.netlify.app%2F&os=iOS&os_version=14.6&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-1020--open-sauced.netlify.app/) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60bbbd7cf5c61a00070fda70&commentID=60bbddcb6deb950b376ad519&resolution=resolved)

**Comment by 0-vortex** [comment]
I have experience using the filter function on monochrome icons, as @filiptronicek mentioned you can use invert to create any color even without the fill on svg. Have some accompanying code for that in my cf-workers repo if needed 🍕

**Comment by mtfoley** [comment]
I use filter: invert(0.85) in index.css with the current changes, and I prob just have more find & replace to run for the remainder of the relevant img tags. At one point I was inverting *all* img tags and the splash screen gave me a good chuckle. But yeah, never had to do anything with it before :-)

**Comment by mtfoley** [comment]
And I forgot docs...

**Comment by mtfoley** [comment]
@filiptronicek @bdougie @0-vortex  I made the updates you asked for here.  

- [x]  For [react-loading-skeleton](https://github.com/dvtng/react-loading-skeleton/blob/master/src/skeleton.js#L18), I added to index.css to handle the color variables in dark mode.  
- [x] From what I found in searching for img tags, I had just the new recommended repos section to add the svg class to.  
- [x] Docs: I added a section toward the end of README.md.

File: src/components/Repository.js [hunk 1]
```diff
@@ -211,7 +211,7 @@ function Repository({user, match}) {
               </div>
             </span>
           ) : (
-            <SkeletonTheme color="#fbfbfb" highlightColor="#f2f2f2">
```
Comment (bdougie) [inline src/components/Repository.js:214]: Thanks for moving this

**Review by bdougie — APPROVED** [review]
This looks good to me. I don't have a lot of context on approaching dark mode technically, but all this makes sense to me. 

My only ask would be adding links to blog posts or guides you used to solve this to the PR body. Mainly because, future contributors will have questions on the decisions made and I want to point to this PR in the future.

**Comment by mtfoley** [comment]
@bdougie I don't think I caught what you asked me to add in that first sentence, other than links to what I referred to.

RE: The general methodology, I couldn't find any exact precedent because some tutorials are written around using prefers-color-scheme to determine an appropriate theme, and others are written around sort of controlling the theme completely within the app in a two-state paradigm, but I didn't see any that gave examples of how to do both (three states like what we have here). Open Sauced might be somewhat unique in that after the PR (maybe?) The rationale of falling back to system preferences was largely based on the StackOverflow [blog](https://stackoverflow.blog/2020/03/31/building-dark-mode-on-stack-overflow/) that @0-vortex pointed out.  I didn't wind up using any of the toggle buttons
RE: img tags with src={svg}, I referred to this [post](https://medium.com/@union_io/swapping-fill-color-on-image-tag-svgs-using-css-filters-fa4818bf7ec6) and [this](https://catalin.red/svg-favicon-light-dark-theme/).
RE: ThemeContext. I used a context because I wanted to define the theme as high up in the component tree as I could, but I wasn't sure how deeply in the component tree I would need to interact with it. [React Context Docs](https://reactjs.org/docs/context.html#consuming-multiple-contexts)

However, even though I use a context, I actually refer to the context very little if at all other than the toggle buttons themselves.  This is because the "system" state of the context is still somewhat indeterminate from any components' perspective of choosing between light and dark.  Because of this, no components actually choose between light and dark - the CSS file and the CSS portion of the styled-components code leverage the `<body class="dark">` to decide how to look.

**Review by filiptronicek — CHANGES_REQUESTED** [review]
Everything looks good, just my background color does not work properly on Firefox...

**Brave**:
image

**Firefox Dev**:
image


Also, can we make the search icon also inverted? Would be a nice touch.
image

**Comment by mtfoley** [comment]
I think I've got firefox fixed now, and I ended up just creating a 2nd version of the search background SVG for dark mode. Still just using body.dark class to differentiate. 

I also made a whole bunch more tweaks to things like the issues icons, form field colors, and rendered note colors.

**Comment by bdougie** [comment]
can confirm that firefox looks better. going to merge. If anything else is caught, we can move it to an issue. Thanks a ton for taking this on @mtfoley and thanks for testing other browsers @filiptronicek. I got to get out of my chrome for everyone's mindset.  

Screen Shot 2021-06-09 at 7 04 03 PM

**Pull request merged**